lib/rubygems/remote_fetcher.rb

```diff
@@ -21,19 +22,24 @@ class Gem::RemoteFetcher
   class FetchError < Gem::Exception
     ##
     # The URI which was being accessed when the exception happened.
+    def self.build(message, uri)
+      original_uri = uri.dup
+      uri = Gem::PrintableUri.parse_uri(uri)
 
-    attr_accessor :uri, :original_uri
-
-    def initialize(message, uri)
-      super message
+      if uri.respond_to?(:original_password) && uri.original_password
+        message = message.sub(uri.original_password, 'REDACTED')
+      end
 
-      uri = Gem::UriParser.parse_uri(uri)
+      new(message, uri.to_s, original_uri)
+    end
 
-      @original_uri = uri.dup
+    attr_accessor :uri, :original_uri
 
-      uri.password = 'REDACTED' if uri.respond_to?(:password) && uri.password
+    def initialize(message, uri, original_uri = nil)
```

Comment:
The `Gem::RemoteFetcher::FetchError` is used in bundler too and if we break the class initialization signature, those tests for checking bundler backward compatibility with old RubyGems will break. 

The `original_uri` use to be generated inside the initializer.

Reply:
See the previous comment.